### PR TITLE
Localization for hour cycles

### PR DESCRIPTION
**Pull request by alexzhirkevich**
## Proposed Changes

Implemented localized `is24HourFormat`.

`TimePicker`s layout now depends on 12/24 hour format of the current locale

## Testing

Test: `PlatformDateFormatTest`

### CODE REVIEW

**Comment by elijah-semyonov** [comment]
Thanks for the PR! 

@eymar 
Can you please have a look at web related part?

**Review by elijah-semyonov — APPROVED** [review]
LGTM, let's wait for a second approval

File: compose/material3/material3/src/jsWasmMain/kotlin/androidx/compose/material3/PlatformDateFormat.jsWasm.kt
Comment (eymar) [inline compose/material3/material3/src/jsWasmMain/kotlin/androidx/compose/material3/PlatformDateFormat.jsWasm.kt]: Firefox seems to not support `hoursCycles`. It doesn't support the method either - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/getHourCycles#supported_hour_cycle_types

It supports only `hourCycle` - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/Locale/hourCycle

But `hourCycle` is defined only when it  was  explcitly declared in Locale constructor. At the same time `hoursCycles` returns an array `['h12']` for `Intl.Locale("en-US")`.

So we need to add more conditioncs. Maybe use `hoursCycles` when it's possible. If not, then `hourCycle`. And if it's undefined, then return false (I guess we have no other option).

As I understand, the current code would crash on firefox. Did you try to run the test on firefox?

Comment (alexzhirkevich) [reply compose/material3/material3/src/jsWasmMain/kotlin/androidx/compose/material3/PlatformDateFormat.jsWasm.kt]: Yep, it would crash, as well as `weekInfo`. Added more conditions and exception checks. 

> But `hourCycle` is defined only when it was explcitly declared in Locale constructor.

So for `hourCycle` we need to use something other than `.toLanguageTag`. But we don't have hourCycle property to declare it in constructor.  I probably can do the same trick as for `firstDayOfWeek` fallback

Comment (alexzhirkevich) [reply compose/material3/material3/src/jsWasmMain/kotlin/androidx/compose/material3/PlatformDateFormat.jsWasm.kt]: Added fallback method

Comment (eymar) [reply compose/material3/material3/src/jsWasmMain/kotlin/androidx/compose/material3/PlatformDateFormat.jsWasm.kt]: Thanks!

**Review by eymar — APPROVED** [review]

File: compose/material3/material3/src/jsWasmMain/kotlin/androidx/compose/material3/PlatformDateFormat.jsWasm.kt
Comment (alexzhirkevich) [reply compose/material3/material3/src/jsWasmMain/kotlin/androidx/compose/material3/PlatformDateFormat.jsWasm.kt]: Cleaned up a little bit, removed `isIntlSupported` check as we catch any exception anyway

File: compose/material3/material3/src/darwinMain/kotlin/androidx/compose/material3/PlatformDateFormat.darwin.kt
Comment (dima-avdeev-jb) [inline compose/material3/material3/src/darwinMain/kotlin/androidx/compose/material3/PlatformDateFormat.darwin.kt]: What means `j` and `a` ?

Comment (alexzhirkevich) [reply compose/material3/material3/src/darwinMain/kotlin/androidx/compose/material3/PlatformDateFormat.darwin.kt]: According to [Unicode LDML](http://www.unicode.org/reports/tr35/tr35-31/tr35-dates.html#Date_Format_Patterns) `j` is 
>This is a special-purpose symbol. It must not occur in pattern or skeleton data. Instead, it is reserved for use in skeletons passed to APIs doing flexible date pattern generation. In such a context, it requests the preferred hour format for the locale (h, H, K, or k), as determined by whether h, H, K, or k is used in the standard short time format for the locale. In the implementation of such an API, 'j' must be replaced by h, H, K, or k before beginning a match against availableFormats data. <ins>Note that use of 'j' in a skeleton passed to an API is the only way to have a skeleton request a locale's preferred time cycle type (12-hour or 24-hour)</ins>.

And `a` is a pattern for `AM` \ `PM` symbol (so presense of this symbol means that locale has 12h format)

Comment (dima-avdeev-jb) [reply compose/material3/material3/src/darwinMain/kotlin/androidx/compose/material3/PlatformDateFormat.darwin.kt]: Thanks! Can you please add a comment about this in code?

Comment (alexzhirkevich) [reply compose/material3/material3/src/darwinMain/kotlin/androidx/compose/material3/PlatformDateFormat.darwin.kt]: Added comment and more test cases to ensure it works correctly

**Review by dima-avdeev-jb — APPROVED** [review]

**Pull request merged**
